File: watcher/util/log.ts
import type { Logger } from "../ports.ts";

function emit(
  level: "info" | "warn" | "error",
  message: string,
  fields?: Record<string, unknown>,
): void {
  const line = JSON.stringify({
    time: new Date().toISOString(),
    level,
    message,
    ...fields,
  });
  if (level === "error") {
    console.error(line);
  } else {
    console.log(line);
  }
}

export function createConsoleLogger(): Logger {
  return {
    info: (message, fields) => emit("info", message, fields),
    warn: (message, fields) => emit("warn", message, fields),
    error: (message, fields) => emit("error", message, fields),
  };
}

File: src-tauri/src/log.rs
use tracing::{error, info, warn};

use crate::ports::Logger;

pub struct ConsoleLogger;

impl Logger for ConsoleLogger {
    fn info(&self, message: &str) {
        info!(target: "scan_carburant", "{message}");
    }

    fn warn(&self, message: &str) {
        warn!(target: "scan_carburant", "{message}");
    }

    fn error(&self, message: &str) {
        error!(target: "scan_carburant", "{message}");
    }
}
